cocos/3d/CCTerrain.cpp

```diff
@@ -490,80 +491,62 @@ cocos2d::Vec3 Terrain::getNormal(int pixel_x, int pixel_y) const
 
 cocos2d::Vec3 Terrain::getIntersectionPoint(const Ray & ray) const
 {
-    Vec3 dir = ray._direction;
-    dir.normalize();
-    Vec3 rayStep = _terrainData._chunkSize.width*0.25*dir;
-    Vec3 rayPos =  ray._origin;
-    Vec3 rayStartPosition = ray._origin;
-    Vec3 lastRayPosition =rayPos;
-    rayPos += rayStep; 
-    // Linear search - Loop until find a point inside and outside the terrain Vector3 
-    Vec3 normal;
-    float height = getHeight(rayPos.x, rayPos.z, &normal);
-    while (rayPos.y > height)
-    {
-        lastRayPosition = rayPos; 
-        rayPos += rayStep; 
-        if (normal.isZero())
-            return Vec3(0, 0, 0);
-        height = getHeight(rayPos.x,rayPos.z); 
-    } 
-
-    Vec3 startPosition = lastRayPosition;
-    Vec3 endPosition = rayPos;
-
-    for (int i= 0; i< 32; i++) 
-    { 
-        // Binary search pass 
-        Vec3 middlePoint = (startPosition + endPosition) * 0.5f;
-        if (middlePoint.y < height) 
-            endPosition = middlePoint; 
-        else 
-            startPosition = middlePoint;
-    } 
-    Vec3 collisionPoint = (startPosition + endPosition) * 0.5f; 
-    return collisionPoint;
+    Vec3 collisionPoint;
+    if (getIntersectionPoint(ray, collisionPoint))
+    {
+        return collisionPoint;
+    }
+    else
+    {
+        return Vec3(0,0,0);
```

Comment:
cocos2d::Vec3 getIntersectionPoint(const Ray & ray) const should be deprecated.
Use bool getIntersectionPoint(const Ray & ray, Vec3 & intersectionPoint) const instead.

There is no correct value to return when there is no intersection point.